File: scan_pending.js
const fs = require('fs');

// 读取并解析 futures_data.js
const content = fs.readFileSync('./futures_data.js', 'utf-8');
const match = content.match(/const FUTURES_DATA = ({[\s\S]*?});/);

if (!match) {
    console.log("ERROR: 找不到 FUTURES_DATA");
    process.exit(1);
}

const FUTURES_DATA = eval('(' + match[1] + ')');

console.log("=".repeat(80));
console.log("📊 扫描所有期货商品 - 3根K线 Pending 形态分析");
console.log("=".repeat(80));
console.log();

const pendingLong = [];
const pendingShort = [];
const activeLong = [];
const activeShort = [];

for (const [code, future] of Object.entries(FUTURES_DATA)) {
    if (!future.main || !future.main.data || future.main.data.length < 3) {
        continue;
    }

    const bars = future.main.data;
    const w1 = bars[bars.length - 3];  // 起点
    const w2 = bars[bars.length - 2];  // Peak/Trough
    const w3 = bars[bars.length - 1];  // Current

    const name = future.name || code;
    const symbol = future.main.symbol || 'N/A';

    // Pending Long规则 (4个条件)
    const cond1Long = w2.high > w1.high;
    const cond2Long = w3.close > w1.low;
    const cond3Long = w3.close <= w2.high;
    const cond4Long = w3.high < w2.high;  // 新增: w3确实回调
    const isPendingLong = cond1Long && cond2Long && cond3Long && cond4Long;

    // Active Long
    const isActiveLong = (w2.high > w1.high) && (w3.close > w2.high);

    // Pending Short规则 (4个条件)
    const cond1Short = w2.low < w1.low;
    const cond2Short = w3.close < w1.high;
    const cond3Short = w3.close >= w2.low;
    const cond4Short = w3.low > w2.low;  // 新增: w3确实反弹
    const isPendingShort = cond1Short && cond2Short && cond3Short && cond4Short;

    // Active Short
    const isActiveShort = (w2.low < w1.low) && (w3.close < w2.low);

    if (isPendingLong) {
        pendingLong.push({
            code, name, symbol,
            w1High: w1.high, w1Low: w1.low,
            w2High: w2.high,
            w3Close: w3.close,
            distance: ((w2.high - w3.close) / w2.high * 100).toFixed(2)
        });
    }

    if (isActiveLong) {
        activeLong.push({ code, name, symbol, close: w3.close, breakout: w2.high });
    }

    if (isPendingShort) {
        pendingShort.push({
            code, name, symbol,
            w1High: w1.high, w1Low: w1.low,
            w2Low: w2.low,
            w3Close: w3.close,
            distance: ((w3.close - w2.low) / w2.low * 100).toFixed(2)
        });
    }

    if (isActiveShort) {
        activeShort.push({ code, name, symbol, close: w3.close, breakdown: w2.low });
    }
}

// 输出结果
console.log(`🟡 Pending Long (蓄势做多) - 共 ${pendingLong.length} 个品种`);
console.log("-".repeat(80));
pendingLong.forEach(item => {
    console.log(`✅ ${item.name} (${item.code}) - ${item.symbol}`);
    console.log(`   w1: High ${item.w1High}, Low ${item.w1Low}`);
    console.log(`   w2: High ${item.w2High} ← 阻力位 (距离${item.distance}%)`);
    console.log(`   w3: Close ${item.w3Close} (蓄势中)`);
    console.log();
});

console.log();
console.log(`🟢 Active Long (已突破) - 共 ${activeLong.length} 个品种`);
console.log("-".repeat(80));
activeLong.forEach(item => {
    console.log(`🚀 ${item.name} (${item.code}) - ${item.symbol}`);
    console.log(`   当前: ${item.close}, 已突破: ${item.breakout}`);
    console.log();
});

console.log();
console.log(`🔴 Pending Short (蓄势做空) - 共 ${pendingShort.length} 个品种`);
console.log("-".repeat(80));
pendingShort.forEach(item => {
    console.log(`⚠️ ${item.name} (${item.code}) - ${item.symbol}`);
    console.log(`   w1: High ${item.w1High}, Low ${item.w1Low}`);
    console.log(`   w2: Low ${item.w2Low} ← 支撑位 (距离${item.distance}%)`);
    console.log(`   w3: Close ${item.w3Close} (蓄势中)`);
    console.log();
});

console.log();
console.log("=".repeat(80));
console.log("总结:");
console.log(`  Pending Long: ${pendingLong.length} 个`);
console.log(`  Active Long: ${activeLong.length} 个`);
console.log(`  Pending Short: ${pendingShort.length} 个`);
console.log(`  Active Short: ${activeShort.length} 个`);
console.log("=".repeat(80));

console.log("\n📋 需要添加 pending_long 标记的商品:");
console.log(pendingLong.map(x => `${x.name}(${x.code})`).join(", "));

console.log("\n📋 需要添加 pending_short 标记的商品:");
console.log(pendingShort.map(x => `${x.name}(${x.code})`).join(", "));

File: scan_pending.py
import json
import re

def analyze_pending_patterns():
    """
    扫描所有期货商品,找出符合3根K线Pending规则的品种
    
    规则 (Pending Long):
    1. w2.high > w1.high  # 有明显上涨
    2. w3.close > w1.low  # 未跌破支撑
    3. w3.close <= w2.high  # 尚未突破
    """
    
    # 读取 futures_data.js
    with open("futures_data.js", "r", encoding="utf-8") as f:
        content = f.read()
    
    # 提取 FUTURES_DATA
    match = re.search(r'const FUTURES_DATA = ({.*?});', content, re.DOTALL)
    if not match:
        print("ERROR: 找不到 FUTURES_DATA")
        return
    
    try:
        data = json.loads(match.group(1))
    except Exception as e:
        print(f"ERROR: 解析JSON失败: {e}")
        return
    
    print("=" * 80)
    print("📊 扫描所有期货商品 - 3根K线 Pending 形态分析")
    print("=" * 80)
    print()
    
    pending_long_list = []
    pending_short_list = []
    active_long_list = []
    active_short_list = []
    
    for code, future in data.items():
        if not future.get('main'):
            continue
        
        main = future['main']
        if not main.get('data') or len(main['data']) < 3:
            continue
        
        # 获取最后3根K线
        bars = main['data']
        w1 = bars[-3]  # 起点
        w2 = bars[-2]  # Peak/Trough
        w3 = bars[-1]  # Current
        
        name = future.get('name', code)
        symbol = main.get('symbol', 'N/A')
        
        # 检查 Pending Long
        cond1_long = w2['high'] > w1['high']
        cond2_long = w3['close'] > w1['low']
        cond3_long = w3['close'] <= w2['high']
        
        is_pending_long = cond1_long and cond2_long and cond3_long
        
        # 检查 Active Long (已突破)
        is_active_long = (w2['high'] > w1['high']) and (w3['close'] > w2['high'])
        
        # 检查 Pending Short
        cond1_short = w2['low'] < w1['low']
        cond2_short = w3['close'] < w1['high']
        cond3_short = w3['close'] >= w2['low']
        
        is_pending_short = cond1_short and cond2_short and cond3_short
        
        # 检查 Active Short (已破位)
        is_active_short = (w2['low'] < w1['low']) and (w3['close'] < w2['low'])
        
        if is_pending_long:
            pending_long_list.append({
                'code': code,
                'name': name,
                'symbol': symbol,
                'w1': w1,
                'w2': w2,
                'w3': w3,
                'resistance': w2['high'],
                'support': w1['low']
            })
        
        if is_active_long:
            active_long_list.append({
                'code': code,
                'name': name,
                'symbol': symbol,
                'w3_close': w3['close'],
                'breakout_level': w2['high']
            })
        
        if is_pending_short:
            pending_short_list.append({
                'code': code,
                'name': name,
                'symbol': symbol,
                'w1': w1,
                'w2': w2,
                'w3': w3,
                'support': w2['low'],
                'resistance': w1['high']
            })
        
        if is_active_short:
            active_short_list.append({
                'code': code,
                'name': name,
                'symbol': symbol,
                'w3_close': w3['close'],
                'breakdown_level': w2['low']
            })
    
    # 输出结果
    print("🟡 Pending Long (蓄势做多) - 共 {} 个品种".format(len(pending_long_list)))
    print("-" * 80)
    for item in pending_long_list:
        print(f"✅ {item['name']} ({item['code']}) - {item['symbol']}")
        print(f"   w1: High {item['w1']['high']}, Low {item['w1']['low']}")
        print(f"   w2: High {item['w2']['high']} ← 阻力位")
        print(f"   w3: Close {item['w3']['close']} (蓄势中)")
        print(f"   突破位: {item['resistance']}, 支撑位: {item['support']}")
        print()
    
    print()
    print("🟢 Active Long (已突破) - 共 {} 个品种".format(len(active_long_list)))
    print("-" * 80)
    for item in active_long_list:
        print(f"🚀 {item['name']} ({item['code']}) - {item['symbol']}")
        print(f"   当前: {item['w3_close']}, 已突破: {item['breakout_level']}")
        print()
    
    print()
    print("🔴 Pending Short (蓄势做空) - 共 {} 个品种".format(len(pending_short_list)))
    print("-" * 80)
    for item in pending_short_list:
        print(f"⚠️ {item['name']} ({item['code']}) - {item['symbol']}")
        print(f"   w1: High {item['w1']['high']}, Low {item['w1']['low']}")
        print(f"   w2: Low {item['w2']['low']} ← 支撑位")
        print(f"   w3: Close {item['w3']['close']} (蓄势中)")
        print(f"   破位: {item['support']}, 阻力位: {item['resistance']}")
        print()
    
    print()
    print("🔻 Active Short (已破位) - 共 {} 个品种".format(len(active_short_list)))
    print("-" * 80)
    for item in active_short_list:
        print(f"📉 {item['name']} ({item['code']}) - {item['symbol']}")
        print(f"   当前: {item['w3_close']}, 已破位: {item['breakdown_level']}")
        print()
    
    print("=" * 80)
    print("总结:")
    print(f"  Pending Long: {len(pending_long_list)} 个")
    print(f"  Active Long: {len(active_long_list)} 个")
    print(f"  Pending Short: {len(pending_short_list)} 个")
    print(f"  Active Short: {len(active_short_list)} 个")
    print("=" * 80)
    
    # 返回需要添加标记的商品代码
    return [item['code'] for item in pending_long_list], \
           [item['code'] for item in pending_short_list]

if __name__ == "__main__":
    pending_long_codes, pending_short_codes = analyze_pending_patterns()
    
    print("\n需要添加 pending_long 标记的商品:")
    print(", ".join(pending_long_codes))
    
    print("\n需要添加 pending_short 标记的商品:")
    print(", ".join(pending_short_codes))
